Feature/primary to maintenance

DimCitus opened the pull request:
Allow `pg_autoctl enable maintenance` on the primary node.

JelteF commented on src/monitor/node_active_protocol.c:
```diff
@@ -1137,76 +1135,109 @@ start_maintenance(PG_FUNCTION_ARGS)
+		SetNodeGoalState(currentNode->nodeName, currentNode->nodePort,
+						 REPLICATION_STATE_MAINTENANCE);
+
+		NotifyStateChange(currentNode->reportedState,
```
Unrelated (so probably better in separate PR), but could this function be refactored to take an `AutoFailoverNode`? Instead of all the fields of it separately. Currently it is hard to review that the right arguments have been passed in.

Or even better, call this from `SetNodeGoalState` (which should then take an `AutoFailoverNode`). So calling it cannot be forgotten.

DimCitus replied:
Good point, and agreed: let's see about that in another PR.

DimCitus replied:
Maybe we should also switch the notification format (and parser) to using JSON by the way. We now have a JSON parser in our code base.

JelteF commented on tests/test_basic_operation.py:
```diff
@@ -53,50 +53,71 @@ def test_005_read_from_secondary():
     node2.run_sql_query("INSERT INTO t1 VALUES (3)")
 
-def test_007_maintenance():
+def test_007_maintenance_primary():
```
Maybe good to add a test to show that the `enable_maintenance` fails without `allowFailover`, so with `@raises(Exception)`.

DimCitus commented on src/bin/pg_autoctl/fsm.c:
```diff
@@ -168,14 +172,15 @@ KeeperFSMTransition KeeperFSM[] = {
 	{ DRAINING_STATE, DEMOTED_STATE, COMMENT_DRAINING_TO_DEMOTED, &fsm_stop_postgres },
 	{ PRIMARY_STATE, DEMOTED_STATE, COMMENT_PRIMARY_TO_DEMOTED, &fsm_stop_postgres },
 	{ PRIMARY_STATE, DEMOTE_TIMEOUT_STATE, COMMENT_PRIMARY_TO_DEMOTED, &fsm_stop_postgres },
+	{ PRIMARY_STATE, MAINTENANCE_STATE, COMMENT_PRIMARY_TO_MAINTENANCE, &fsm_stop_postgres },
```
I'm not sure how to prevent that... other than maybe going from maintenance to demote timeout. In that case pg_autoctl will ensure Postgres is shut down for as long as it runs, at least. Another way to do that might be to prepare the replication setup so that if Postgres is started, it's already setup to be a standby, already in the MAINTENANCE state.

DimCitus replied:
So I think we can still code that without extra states. As we do a clean shutdown on the primary before promoting the secondary, we don't expect to have to do a `pg_rewind` later. Given that we can share the same code for MAINTENANCE to CATCHINGUP in the case when the node used to be a primary or a secondary before entering MAINTENANCE.

If review shows that this is an ill advised position to take, then we will need to introduce a specific primary only maintenance state.

DimCitus commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -692,6 +692,25 @@ bool
+		 */
+		for (int i = 0; i < 3; i++)
+		{
+			if (!pgsql_checkpoint(pgsql))
```
When Postgres stops it does a checkpoint anyway. See  comments in https://github.com/postgres/postgres/blob/master/src/backend/postmaster/checkpointer.c#L13 and here:

```c
 * The checkpointer is started by the postmaster as soon as the startup
 * subprocess finishes, or as soon as recovery begins if we are doing archive
 * recovery.  It remains alive until the postmaster commands it to terminate.
 * Normal termination is by SIGUSR2, which instructs the checkpointer to
 * execute a shutdown checkpoint and then exit(0).  (All backends must be
 * stopped before SIGUSR2 is issued!)  Emergency termination is by SIGQUIT;
 * like any backend, the checkpointer will simply abort and exit on SIGQUIT.
```

The advantage of doing a CHECKPOINT ourselves is that we do that while online, and then the shutdown sequence of Postgres can be fast. I'm not completely sold on the loop-of-3 thing, but certainly reducing the time when Postgres has not shut down yet but can't do anything anymore is a good thing.

DimCitus replied:
Changed it to doing 2 checkpoints before stopping Postgres. I also made it so that if we fail the second checkpoint, we redo only that one rather than two of them again.

JelteF replied:
> but certainly reducing the time when Postgres has not shut down yet but can't do anything anymore is a good thing.

I thought so to, but I don't think anymore this is the case for all our shutdowns. For this maintenance case it definitely seems good. However, when the primary notices it's on the losing side of a network partition it must stop accepting writes quickly. Otherwise there can be a split brain. So waiting for a CHECKPOINT to finish (even a single one), before stopping to accept writes seems bad in that case.

I think this makes it complex enough to warrant a separate PR.

JelteF replied:
Resolving this since it's now only called when going to maintenance.

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -300,6 +310,121 @@ cli_disable_secondary(int argc, char **argv)
+ * cli_secondary_getopts parses command line options for the secondary feature,
+ * both during enable and disable. Little verification is performed however the
+ * function will error when no --pgdata or --formation are provided, existance
+ * of either are not verified.
```
Ah yeah, fixed, thanks. Oops.

DimCitus commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -692,6 +692,25 @@ bool
+		 * third one might have still some work to do.
+		 */
+		for (int i = 0; i < 3; i++)
+		{
```
I added a log about that, good point. Note that Postgres is still available during the checkpoints that we do here.

DimCitus commented on src/bin/pg_autoctl/fsm.c:
```diff
@@ -197,6 +202,11 @@ KeeperFSMTransition KeeperFSM[] = {
+	/*
+	 * primary was put to maintenance
+	 */
+	{ SECONDARY_STATE, WAIT_PRIMARY_STATE, COMMENT_PRIMARY_TO_MAINTENANCE, &fsm_promote_standby },
```
Fixed.

DimCitus commented on src/bin/pg_autoctl/monitor.c:
```diff
@@ -2809,20 +2809,52 @@ monitor_wait_until_primary_applied_settings(Monitor *monitor,
  */
 bool
 monitor_wait_until_node_reported_state(Monitor *monitor,
+									   const char *formation,
```
Fixed, thanks.

onderkalaci commented:
@DimCitus could you please clarify the meaning of `maintenance`  mode a little further? When I read the documentation (and the code), [it](https://pg-auto-failover.readthedocs.io/en/latest/operations.html#maintenance-of-a-secondary-node) says that `so that the Postgres server is not doing synchronous replication anymore`. So, as far as I can see, we set  `synchronous_standby_names` to empty string. This makes a lot of sense for secondaries. 

How do we define `maintenance` for the primary? I have some ideas reading through the code, but it sounds useful if you can explain your perspective so that I don't miss some details. Thanks!

JelteF commented:
@DimCitus I think the best way to handle @onderkalaci his comment is updating the docs in this PR too. Since they will indeed need to be updated anyway.

JelteF commented on src/bin/pg_autoctl/primary_standby.c:
```diff
@@ -764,7 +764,7 @@ primary_rewind_to_standby(LocalPostgresServer *postgres)
 	log_info("Rewinding PostgreSQL to follow new primary %s:%d",
 			 primaryNode->host, primaryNode->port);
 
-	if (!pg_ctl_stop(pgSetup->pg_ctl, pgSetup->pgdata))
```
For a separate PR, but because of this I noticed that there are some more `pg_ctl_stop` calls. I think a few of these should be changed as well.

DimCitus commented:
> @DimCitus I think the best way to handle @onderkalaci his comment is updating the docs in this PR too. Since they will indeed need to be updated anyway.

Fair. I have updated the documentation now, and also fixed the failover/switchover coverage that we have, where it was not using the new commands yet. @onderkalaci : does that help answer your questions?

JelteF commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -753,6 +753,103 @@ fsm_stop_postgres(Keeper *keeper)
+		 */
+		log_info("Preparing shutdown for maintenance: CHECKPOINT;");
+
+		/* we may fail to CHECKPOINT and then will retry at next FSM loop */
```
Do we really want this? If checkpoint fails, maybe we simply want to quit without doing one.
I assume it's not common for CHECKPOINT to fail. So in the rare case that it does, might fail continuously if we keep retrying. Then we never reach maintenance state.

DimCitus replied:
CHECKPOINT may fail because File System Is Full, in which case a maintenance operation is needed on the primary. So yeah I think you're right, we should try to checkpoint twice but accept that it may fail, because the reason for the CHECKPOINT failure might also be the reason why an operator wants to put this node in maintenance...

DimCitus replied:
Done that way now (errors are only a warning and we continue forward with the stopping of Postgres).

onderkalaci commented on docs/operations.rst:
```diff
@@ -73,44 +73,127 @@ so that the Postgres server is not doing *synchronous replication* anymore
+  11:53:12 |   2 | localhost |   5002 |            primary |        maintenance
+  11:53:12 |   2 | localhost |   5002 |        maintenance |        maintenance
+
+One the primary has reached the ``maintenance`` state then the secondary is
```
`one` -> `once`?

DimCitus replied:
Done thanks.

JelteF commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -753,6 +753,103 @@ fsm_stop_postgres(Keeper *keeper)
+ * stopping Postgres to ensure a smooth transition.
+ */
+bool
+fsm_stop_postgres_for_maintenance(Keeper *keeper)
```
Should we maybe also call this when going to maintenance as secondary? Otherwise it might be goodbetter to rename the function to make clear that it's only meant for primary maintenance.

DimCitus replied:
I don't believe it's necessary to CHECKPOINT a secondary before going to maintenance. Also, we don't even stop Postgres there.

JelteF commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -753,6 +753,103 @@ fsm_stop_postgres(Keeper *keeper)
+
+
+/*
+ * fsm_stop_postgres_and_setup_standby is used when the primary is put to
+ * maintenance. Not only do we stop Postgres, we also prepare a partial setup
+ * as a secondary.
+ */
+bool
+fsm_stop_postgres_and_setup_standby(Keeper *keeper)
```
The function name and/or comment don't indicate that this does the checkpoints before stopping or should be used for maintenance only.

DimCitus replied:
Added some comments about that.

DimCitus commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -1040,3 +1147,67 @@ fsm_promote_standby(Keeper *keeper)
+ *
+ *    start_postgres
+ * && promote_standby
+ * && add_standby_to_hba
```
Yes most of the work is done in `prepare_replication`.

DimCitus replied:
Comment reworked.

DimCitus commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -1040,3 +1147,67 @@ fsm_promote_standby(Keeper *keeper)
+ * && disable_synchronous_replication
+ * && keeper_drop_replication_slots_for_removed_nodes
+ *
+ * So we reuse fsm_disable_replication() here, rather than copy/pasting the same
+ * bits code in the fsm_promote_standby() function body. If the definition of
+ * the fsm_promote_standby transition ever came to diverge from whatever
+ * fsm_disable_replication() is doing, we'd have to copy/paste and maintain
+ * separate code path.
```
Well we still reuse `fsm_disable_replication()`, it's just not the only thing we do... do you have a rewrite in mind?

DimCitus replied:
I reworked the comment, we are actually copy/pasting from another transition function and now make that clear. I still believe that we should copy/paste rather than invent an intermediate function and call it from two places with an argument for the expected state of the primary, but maybe we could already do that.

DimCitus commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -1040,3 +1147,67 @@ fsm_promote_standby(Keeper *keeper)
+
+
+/*
+ * fsm_promote_standby is used in several situations in the FSM transitions and
```
Fixed, thanks.

onderkalaci commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -1040,3 +1144,56 @@ fsm_promote_standby(Keeper *keeper)
+ * DEMOTE_TIMEOUT_STATE.
+ */
+bool
+fsm_promote_standby_for_primary_maintenance(Keeper *keeper)
```
minor: Could we change `fsm_promote_standby ` to get the target state and make this one:

```
def fsm_promote_standby_for_primary_maintenance()
    
    fsm_promote_standby(keeper, MAINTENANCE_STATE)
```

DimCitus replied:
Nope, or at least, not like that. Look at `fsm.h` where we define that transition function all only take a single argument, that is a pointer to the current Keeper structure.

``` c
/*
 * Each FSM entry is a transition from a current state to the next
 */
typedef bool (*ReachAssignedStateFunction)(Keeper *keeper);

/* defines a possible transition in the FSM */
typedef struct KeeperFSMTransition
{
	NodeState current;
	NodeState assigned;
	const char *comment;
	ReachAssignedStateFunction transitionFunction;
} KeeperFSMTransition;
```

JelteF replied:
If it's really the same except for the other node state. Then we could have something like this:

def fsm_promote_standby_for_primary_maintenance()    
    fsm_promote_standby(keeper, MAINTENANCE_STATE)

def fsm_promote_standby_normal()
    fsm_promote_standby(keeper, DEMOTE_TIMEOUT_STATE)

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -425,8 +559,10 @@ cli_disable_maintenance(int argc, char **argv)
+			keeper.config.groupId,
 			keeper.state.current_node_id,
-			CATCHINGUP_STATE))
+			SECONDARY_STATE))
```
The reason we change that here is because we now display the transitions as they are happening. In the master branch we did not, so we exited as soon as possible. In this PR we think it's more useful to show an extra transition even if that means waiting longer.

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -22,11 +22,13 @@
 #include "parsing.h"
 #include "pgsetup.h"
 
+bool allowFailover = false;
```
Nope. At the moment we have a not ideal situation where the command line option parsing is kept in sync with the configuration file parsing. That's a nice user friendly property, but what do we do about options that are not relevant to the configuration and only apply for the duration of an interactive command?

Well `allowFailover` falls into that bin, and as such we need to have a global variable for it. That's also because of the way we are handling command lines parsing function and `getopt` functions: those can't share state, so we need a global one.

onderkalaci commented on src/monitor/group_state_machine.c:
```diff
@@ -109,14 +109,42 @@ ProceedGroupState(AutoFailoverNode *activeNode)
-				 errdetail("activeNode is %s:%d in state %s",
-						   activeNode->nodeName, activeNode->nodePort,
-						   ReplicationStateGetName(activeNode->goalState))));
+		AutoFailoverNode *otherNode = OtherNodeInGroup(activeNode);
```
(Again, some questions to understand better, not necessarily asking for change at this point)

I think for the first time we end up with `no primary exists` case. Does that mean that we're losing our `service availability` principle for this short duration? 

A semi-related question: If the user follows `perform switchover && enable maintenance on the secondary` pattern to enable maintenance on the original primary, do you see any differences in terms of `service availability`?

DimCitus replied:
In terms of service availability this PR does not change much, we still need to stop the primary before promoting the standby as this is the only way to make sure we're not going to miss writes and have a split-brain situation.

In terms of state machine, the state maintenance could belong to either a primary or a secondary in the coding you reviewed and I agree it makes it harder to reason about. So I just introduced a new state `prepare_promotion` that only belongs to the primary, and I think that solves it.

JelteF commented:
After trying the new command I found a few issues:
Issue 1:
```bash
pg_autoctl enable maintenance --pgdata node2
pg_autoctl enable maintenance --pgdata node2 # blocks forever
```

Issue 2a:
```bash
pg_autoctl enable maintenance --pgdata node2 \
     && pg_autoctl disable maintenance --pgdata node2
# second command fails with the following error:
# ERROR:  cannot stop maintenance when current state for node 172.24.28.118:5433 is "primary"
```
I guess this could be fixed by waiting for two events before succeeding enable maintenance:
1. current node `maintenance`
2. othernode `wait_primary`

Issue 2b:
Same as issue 2a, but with --allow-failover and different state in the error message:
```bash
pg_autoctl enable maintenance --pgdata node1 --allow-failover \
    && pg_autoctl disable maintenance --pgdata node1
# ERROR:  cannot stop maintenance when current state for node 172.24.28.118:5434 is "secondary"
```

Issue 3:
This might not really be an issue, but doing disable+enable maintenance is now faster than `pg_autoctl switchover`. Which seems strange and sounds like we might want to change `switchover` to do the same or at least something similar:
```bash
# sleep 2 is needed because of issue 2b
pg_autoctl enable maintenance --pgdata node1 --allow-failover \
    && sleep 2 && pg_autoctl disable maintenance --pgdata node1
15:03:31 26388 INFO  Listening monitor notifications about state changes in formation "default" and group 0
15:03:31 26388 INFO  Following table displays times when notifications are received
    Time |  ID |          Host |   Port |      Current State |     Assigned State
---------+-----+---------------+--------+--------------------+-------------------
15:03:32 |   1 | 172.24.28.118 |   5433 |            primary |        maintenance
15:03:32 |   1 | 172.24.28.118 |   5433 |        maintenance |        maintenance
15:03:34 26523 INFO  Listening monitor notifications about state changes in formation "default" and group 0
15:03:34 26523 INFO  Following table displays times when notifications are received
    Time |  ID |          Host |   Port |      Current State |     Assigned State
---------+-----+---------------+--------+--------------------+-------------------
15:03:35 |   1 | 172.24.28.118 |   5433 |        maintenance |         catchingup
15:03:35 |   1 | 172.24.28.118 |   5433 |         catchingup |         catchingup
15:03:36 |   1 | 172.24.28.118 |   5433 |         catchingup |          secondary
15:03:36 |   2 | 172.24.28.118 |   5434 |       wait_primary |            primary
15:03:36 |   1 | 172.24.28.118 |   5433 |          secondary |          secondary
```

```
pg_autoctl perform switchover --pgdata node2
15:04:12 28074 INFO  Targetting group 0 in formation "default"
15:04:12 28074 INFO  Listening monitor notifications about state changes in formation "default" and group 0
15:04:12 28074 INFO  Following table displays times when notifications are received
    Time |  ID |          Host |   Port |      Current State |     Assigned State
---------+-----+---------------+--------+--------------------+-------------------
15:04:13 |   2 | 172.24.28.118 |   5434 |            primary |           draining
15:04:13 |   1 | 172.24.28.118 |   5433 |          secondary |  prepare_promotion
15:04:13 |   2 | 172.24.28.118 |   5434 |           draining |           draining
15:04:13 |   1 | 172.24.28.118 |   5433 |  prepare_promotion |  prepare_promotion
15:04:14 |   1 | 172.24.28.118 |   5433 |  prepare_promotion |   stop_replication
15:04:14 |   2 | 172.24.28.118 |   5434 |           draining |     demote_timeout
15:04:15 |   2 | 172.24.28.118 |   5434 |     demote_timeout |     demote_timeout
15:04:15 |   1 | 172.24.28.118 |   5433 |   stop_replication |   stop_replication
15:04:16 |   1 | 172.24.28.118 |   5433 |   stop_replication |       wait_primary
15:04:16 |   2 | 172.24.28.118 |   5434 |     demote_timeout |            demoted
15:04:16 |   1 | 172.24.28.118 |   5433 |       wait_primary |       wait_primary
15:04:17 |   2 | 172.24.28.118 |   5434 |            demoted |            demoted
15:04:18 |   2 | 172.24.28.118 |   5434 |            demoted |         catchingup
15:04:18 |   2 | 172.24.28.118 |   5434 |         catchingup |         catchingup
15:04:19 |   2 | 172.24.28.118 |   5434 |         catchingup |          secondary
15:04:19 |   1 | 172.24.28.118 |   5433 |       wait_primary |            primary
15:04:19 |   2 | 172.24.28.118 |   5434 |          secondary |          secondary
15:04:19 |   1 | 172.24.28.118 |   5433 |            primary |            primary
```

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -61,17 +63,17 @@ static CommandLine disable_secondary_command =
 	make_command("maintenance",
 				 "Disable Postgres maintenance mode on this node",
-				 CLI_PGDATA_USAGE,
+				 " [ --pgdata --allow-failover ]",
```
Remove `--allow-failover` from the command help string, thanks.

DimCitus commented on docs/operations.rst:
```diff
@@ -73,44 +73,127 @@ so that the Postgres server is not doing *synchronous replication* anymore
 
-It is possible to trigger a failover manually with pg_auto_failover, by
-using the SQL API provided by the monitor::
+  $ pg_autoctl disable maintenance
```
A failover/switchover is always going to impact the application, that needs to disconnect and reconnects. Best case, the application implement client-side HA properly and that's good. Anyway, don't failover for the sake of failing over, there's no reason to switch back in that case.

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -133,6 +135,7 @@ cli_secondary_getopts(int argc, char **argv)
 	static struct option long_options[] = {
 		{ "pgdata", required_argument, NULL, 'D' },
 		{ "formation", required_argument, NULL, 'f' },
+		{ "allow-failover", no_argument, NULL, 'A' },
```
1. `getopt` is pretty old and crufty ; fixed now
  2. it's completely different

```
$ pg_autoctl enable
pg_autoctl enable: expected a command

Available commands:
  pg_autoctl enable
    secondary    Enable secondary nodes on a formation
    maintenance  Enable Postgres maintenance mode on this node
    ssl          Enable SSL configuration on this node
```

DimCitus commented on docs/operations.rst:
```diff
@@ -73,44 +73,127 @@ so that the Postgres server is not doing *synchronous replication* anymore
+  11:53:11 50599 INFO  Listening monitor notifications about state changes in formation "default" and group 0
+  11:53:11 50599 INFO  Following table displays times when notifications are received
+      Time |  ID |      Host |   Port |      Current State |     Assigned State
+  ---------+-----+-----------+--------+--------------------+-------------------
```
We filter events to the local node at the moment, we could filter to the group and wait until some other node has reached wait_primary. That should be easy enough to do, but we need another kind of function for that.

DimCitus replied:
Done that way now.

DimCitus commented:
@JelteF I have fixed the infinite wait: that's because `start_maintenance` now returns true when a node is already in the maintenance state. Also, I have now added an intermediate step and this looks like this:

```
$ pg_autoctl enable maintenance --allow-failover
18:11:24 16402 INFO  Listening monitor notifications about state changes in formation "default" and group 0
18:11:24 16402 INFO  Following table displays times when notifications are received
    Time |  ID |      Host |   Port |       Current State |      Assigned State
---------+-----+-----------+--------+---------------------+--------------------
18:11:25 |   1 | localhost |   5001 |             primary | prepare_maintenance
18:11:25 |   2 | localhost |   5002 |           secondary |   prepare_promotion
18:11:25 |   2 | localhost |   5002 |   prepare_promotion |   prepare_promotion
18:11:25 |   1 | localhost |   5001 | prepare_maintenance | prepare_maintenance
18:11:26 |   2 | localhost |   5002 |   prepare_promotion |        wait_primary
18:11:26 |   1 | localhost |   5001 | prepare_maintenance |         maintenance
18:11:27 |   2 | localhost |   5002 |        wait_primary |        wait_primary

$ pg_autoctl disable maintenance
18:11:32 16470 INFO  Listening monitor notifications about state changes in formation "default" and group 0
18:11:32 16470 INFO  Following table displays times when notifications are received
    Time |  ID |      Host |   Port |       Current State |      Assigned State
---------+-----+-----------+--------+---------------------+--------------------
18:11:33 |   1 | localhost |   5001 |         maintenance |          catchingup
18:11:33 |   1 | localhost |   5001 |          catchingup |          catchingup
18:11:34 |   1 | localhost |   5001 |          catchingup |           secondary
18:11:34 |   2 | localhost |   5002 |        wait_primary |             primary
18:11:34 |   1 | localhost |   5001 |           secondary |           secondary
18:11:34 |   2 | localhost |   5002 |             primary |             primary
```

With an intermediate step we have a cleaner code to maintain in the group state machine, and also the over all transitions make more sense, maybe even from a user perspective.

onderkalaci commented on docs/fsm.rst:
```diff
@@ -131,6 +131,14 @@ secondary will be online to accept writes. When the old primary reaches
 the wait_primary state then the secondary is safe to take offline with
 minimal consequences.
 
+**Prepare_maintenance**
```
(unrelated comment)
Wow, I haven't seen `docs/fsm.rst`, good to now. I was actually trying to understand why do we have `demote_timeout` state right now :)

onderkalaci commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -334,17 +459,31 @@ cli_enable_maintenance(int argc, char **argv)
 
-	if (!pgsql_listen(&(keeper.monitor.pgsql), channels))
+	/*
+	 * If we're already in MAINTENANCE we monitor returns true but we don't
```
`we monitor`  ->  `the monitor`?

DimCitus replied:
Fixed, thanks!

DimCitus commented on src/bin/pg_autoctl/fsm.c:
```diff
@@ -125,7 +125,8 @@
 
 #define COMMENT_BLOCKED_WRITES \
-	"Promoting the standby node after the primary has blocked writes"
+	"Promoting a Citus Worker standby after having blocked writes " \
```
Yes indeed, that's the special FSM path taken by a Citus worker behind a Citus coordinator.

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -356,13 +495,17 @@ cli_enable_maintenance(int argc, char **argv)
-			&(keeper.monitor),
-			keeper.state.current_node_id,
-			MAINTENANCE_STATE))
+	if (keeper.state.current_role != MAINTENANCE_STATE)
```
Fixed now, thanks.

DimCitus commented on docs/operations.rst:
```diff
@@ -73,44 +73,136 @@ so that the Postgres server is not doing *synchronous replication* anymore
+  13:13:47 |   2 | localhost |   5002 |         maintenance |         maintenance
+
+Once the primary has reached the ``maintenance`` state then the secondary is
+promoted, which we don't see here. When the operation is done we can have
```
Fixed.

DimCitus commented on src/bin/pg_autoctl/fsm_transition.c:
```diff
@@ -753,6 +753,100 @@ fsm_stop_postgres(Keeper *keeper)
+ *
+ * Also, to be user friendly, we try to CHECKPOINT while the read-write traffic
+ * is still open to the primary server, in order to prepare a faster Postgres
+ * shutdown.
```
Fixed.

DimCitus commented on src/bin/pg_autoctl/monitor.c:
```diff
@@ -2933,9 +2807,10 @@ monitor_wait_until_node_reported_state(Monitor *monitor,
-monitor_wait_until_new_primary(Monitor *monitor,
-							   const char *formation,
-							   int groupId)
+monitor_wait_until_some_node_reported_state(Monitor *monitor,
```
Fixed.

JelteF commented on docs/fsm.rst:
```diff
@@ -131,6 +131,14 @@ secondary will be online to accept writes. When the old primary reaches
 the wait_primary state then the secondary is safe to take offline with
 minimal consequences.
 
+**Prepare_maintenance**
```
`wait_maintenance` state should alsoget a description here. While reading through the list I also realised that `prepare_promotion` is not in there either. Probably good to add it.

DimCitus replied:
Added, thanks!

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -300,6 +310,121 @@ cli_disable_secondary(int argc, char **argv)
+
+	static struct option long_options[] = {
+		{ "pgdata", required_argument, NULL, 'D' },
+		{ "allow-failover", no_argument, NULL, 'A' },
```
Yeah https://www.man7.org/linux/man-pages/man3/getopt.3.html is a strange beast. I have in mind that someday we might want to see about using https://www.gnu.org/software/libc/manual/html_node/Argp.html#Argp or even implementing our own version of it, but that's too much work without a clear outcome at the moment.

DimCitus commented on src/bin/pg_autoctl/cli_enable_disable.c:
```diff
@@ -334,17 +459,31 @@ cli_enable_maintenance(int argc, char **argv)
 		exit(EXIT_CODE_KEEPER);
 	}
 
+	if (keeper.state.current_role == PRIMARY_STATE && !allowFailover)
```
Indeed your understanding is correct. We need to think of `pg_autoctl` itself as a client-server program/architecture. We have a Unix service (or daemon) running in the background, and some short lived commands that run in the foreground and have impact on the background process, either sending it a signal (`pg_autoctl reload`, `pg_autoctl stop`) or interacting with the monitor directly (`pg_autoctl perform failover`, `pg_autoctl enable maintenance`).

DimCitus commented on docs/operations.rst:
```diff
@@ -73,44 +73,139 @@ so that the Postgres server is not doing *synchronous replication* anymore
 
-It is possible to trigger a failover manually with pg_auto_failover, by
-using the SQL API provided by the monitor::
+The same command ``pg_autoctl enable maintenance`` implements that operation
```
1. we should discard `--disable-monitor` entirely at the moment, just make the code not assume we have a monitor to talk to ; it's a WIP patch that never made it fully, maybe should be ripped but I'm not sure if that would not be more work that what we do now
  2. yeah it's expected, if `pg_autoctl` is not working (as a service, in the background, when in production, usually) then no progress can be made ; I think that having the user wait is a good sign to remind them to run the process ; also if `pg_autoctl` is not running neither is Postgres, so...
